pinot-query-planner/src/main/java/org/apache/pinot/query/planner/physical/SmartShuffleRewriteVisitor.java

```diff
@@ -0,0 +1,468 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.calcite.rel.RelDistribution;
+import org.apache.pinot.common.config.provider.TableCache;
+import org.apache.pinot.core.transport.ServerInstance;
+import org.apache.pinot.query.planner.QueryPlan;
+import org.apache.pinot.query.planner.StageMetadata;
+import org.apache.pinot.query.planner.logical.RexExpression;
+import org.apache.pinot.query.planner.logical.StageLayoutVisitor;
+import org.apache.pinot.query.planner.logical.StageLayoutVisitor.StageLayout;
+import org.apache.pinot.query.planner.partitioning.FieldSelectionKeySelector;
+import org.apache.pinot.query.planner.partitioning.KeySelector;
+import org.apache.pinot.query.planner.stage.AggregateNode;
+import org.apache.pinot.query.planner.stage.FilterNode;
+import org.apache.pinot.query.planner.stage.JoinNode;
+import org.apache.pinot.query.planner.stage.MailboxReceiveNode;
+import org.apache.pinot.query.planner.stage.MailboxSendNode;
+import org.apache.pinot.query.planner.stage.ProjectNode;
+import org.apache.pinot.query.planner.stage.SortNode;
+import org.apache.pinot.query.planner.stage.StageNode;
+import org.apache.pinot.query.planner.stage.StageNodeVisitor;
+import org.apache.pinot.query.planner.stage.TableScanNode;
+import org.apache.pinot.query.planner.stage.ValueNode;
+import org.apache.pinot.spi.config.table.ColumnPartitionConfig;
+import org.apache.pinot.spi.config.table.IndexingConfig;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This shuffle optimizer uses a {@link Set<PartitionKey>} to keep track of partition-keys for each inner stage node.
+ * If there are two partition keys P1 and P2 in the set for a given inner stage node, that means the data for that
+ * stage is partitioned by both P1 and P2 (independently). An example instance where this can happen is a join stage.
+ * In the join-stage, the data in that stage is partitioned by both a partition-key on the left and a partition-key
+ * on the right.
+ *
+ * If a sending stage has the keys {P1, P2} and the receiving stage needs data partitioned by P1, then a shuffle is
+ * not needed. Moreover, if P1 = [0, 1] and the keys are again {P1, P2} (value of P2 is arbitrary), if the receiving
+ * stage needs data partitioned by [0, 1, 2], then again no shuffle is needed.
+ *
+ * Also see: {@link PartitionKey} for its definition.
+ * TODO: Use a smarter and more appropriate name than "SmartShuffleRewriterVisitor".
+ */
+public class SmartShuffleRewriteVisitor
+    implements StageNodeVisitor<Set<SmartShuffleRewriteVisitor.PartitionKey>, StageLayout> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SmartShuffleRewriteVisitor.class);
+
+  private final TableCache _tableCache;
+  private final Map<Integer, StageMetadata> _stageMetadataMap;
+  private boolean _canSkipShuffleForJoin;
```

Comment:
IIUC, the idea here is that,
1. there are some additional metadata that needs to be computed to run the colocated-join smart shuffle visitor. (e.g. the StageLayout context)
2. then the smartShuffleVisitor is run and produces something locally private to the visitor itself. 

with these 2 assumptions. we can probably make WorkerAssignmentVisitor as an SPI and load them dynamically based on query option. 

I am thinking of something like this:


QueryEnvironment adds 
```
_workerAssignmentStrategyFactory;
runWorkerAssignmentStrategy(QueryPlan, WorkerManager);
```

added class `WorkerAssignmentStrategyFactory` with
```
WorkerAssignmentStrategy strategy = WorkerAssignmentStrategyFactory.createStrategy(QueryPlan, WorkerManager);
```
 to create different worker assignment strategy based on QueryPlan

another class `WorkerAssignmentStrategy` that
```
WorkerAssignmentStrategy<Context, Layout> implements StageNodeVisitor<Context, Layout>
```